src/main/java/com/aws/greengrass/lifecyclemanager/KernelLifecycle.java

```diff
@@ -457,11 +457,24 @@ public void startupAllServices() {
      * Shutdown all services in dependency order.
      *
      * @param timeoutSeconds timeout seconds for waiting all services to shutdown. Use -1 to wait infinitely.
+     * @param serviceNames Configuration of services yet to be merged
      */
     @SuppressWarnings("PMD.AvoidCatchingThrowable")
-    public void stopAllServices(int timeoutSeconds) {
-        GreengrassService[] d = kernel.orderedDependencies().toArray(new GreengrassService[0]);
+    public void stopAllServices(int timeoutSeconds, Set<String> serviceNames) {
+        HashSet<String> runningServices = kernel.orderedDependencies().stream().map(GreengrassService::getName)
+                .collect(Collectors.toCollection(HashSet::new));
+        ArrayList<GreengrassService> newServices = serviceNames.stream()
+                .filter(service -> !runningServices.contains(service))
+                .collect(Collectors.toCollection(ArrayList::new))
+                .stream()
+                .map(kernel::locateIgnoreError)
+                .collect(Collectors.toCollection(ArrayList::new));
+        logger.atDebug()
+                .kv("new-services", newServices.stream().map(GreengrassService::getName)
+                        .collect(Collectors.toCollection(ArrayList::new))).log();
```

Comment:
You don't seem to be using the newServices list for anything?

Reply:
Yes, tbh, I just need the debug log to avoid PMD violation
locateIgnoreError should create the new services which do not exist and the original loop to close all services should close the newly created ones. This is able to resolve the scenario where a deployment consists of
1. Nucleus OTA update
2. Component with long install lifecycle

Reply:
Then I don't think you need to do this locate stuff in the `stopAllServices` method, you could instead make another method called `initServices` or something with the newConfig object and call it before `stopAllServices`. If you plan to keep it this way then definitely add comments explaining what's being done.

Reply:
Secondly, in that case you only need to call `locate` on the service list from newConfig because like I said, locate doesn't do anything but return the existing instance for a running service. And you can use a simple forEach loop for that

Reply:
Right, but `locate` will throw an error on services which do not exist, `localIgnoreError` will create the service if it does not exist. I think creating the new service and closing that should silence the dependency init in the lifecycle listener and new services will  get re-installed after the bootstrap steps are completed.

Reply:
I mean, both `locate` and `locateIgnoreError` will return the existing instance when called on a running service, so you don't need to filter running services out from the list you get from `newConfig`, that's all

Reply:
I see, yeah that makes sense